File: solution-python/017_letter_combinations_of_a_phone_number.py
"""
	017. Letter Combinations of a Phone Number
	@name: li jin
	@date: Feb 28, 2016
	@link: https://leetcode.com/problems/letter-combinations-of-a-phone-number/
	@time: 44 ms
"""

class Solution(object):
    def letterCombinations(self, digits):
        letters = [" ", "*", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        ans = []
        for d in digits:
            d = int(d)
            if ans:
                tmp = []
                for t in letters[d]:
                    tmp += [a + t for a in ans]
                ans = tmp
            else:
                ans = [t for t in letters[d]]
        return ans

File: solution-cpp/017_letter_combinations_of_a_phone_number.cpp
/*
	017. Letter Combinations of a Phone Number
	@name: li jin
	@date: Feb 28, 2016
	@link: https://leetcode.com/problems/letter-combinations-of-a-phone-number/
	@time: 0 ms
*/

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> ans;
        string letters[10] = {" ", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        for (auto d : digits) {
            if (ans.size() > 0) {
                vector<string> tmp;
                for (auto c : letters[d - '0']) {
                    for (auto s : ans) {
                        tmp.push_back(s + c);
                    }
                }
                ans = tmp;
            }
            else {
                for (auto c : letters[d - '0']) {
                    stringstream ss;
                    ss << c;
                    ans.push_back(ss.str());
                }
            }
        }
        return ans;
    }
};